File: src/treemapper/diffctx/git.py
from __future__ import annotations

import logging
import re
import subprocess
from pathlib import Path
from types import TracebackType

from .types import DiffHunk

logger = logging.getLogger(__name__)

_HUNK_RE = re.compile(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@")
_RANGE_RE = re.compile(r"^\s*(\S+?)(\.\.\.?)(\S*?)\s*$")  # NOSONAR(S5852)
_SAFE_RANGE_RE = re.compile(r"^[a-zA-Z0-9_.^~/@{}\-]+(\.\.\.?[a-zA-Z0-9_.^~/@{}\-]*)?$")
_SAFE_DIFF_FLAGS = ["--no-textconv", "--no-ext-diff"]


class GitError(Exception):
    pass


def _validate_diff_range(diff_range: str) -> None:
    if not _SAFE_RANGE_RE.match(diff_range.strip()):
        raise GitError(f"Invalid diff range: {diff_range!r}")


def run_git(repo_root: Path, args: list[str]) -> str:
    try:
        result = subprocess.run(
            ["git", "-C", str(repo_root), *args],
            capture_output=True,
            encoding="utf-8",
            errors="replace",
            check=True,
            timeout=60,
        )
        return result.stdout
    except subprocess.CalledProcessError as e:
        raise GitError(f"git {' '.join(args)} failed: {e.stderr.strip()}") from e
    except subprocess.TimeoutExpired as e:
        raise GitError(f"git {' '.join(args)} timed out after 60s") from e
    except FileNotFoundError as e:
        raise GitError("git is not installed or not in PATH") from e


def is_git_repo(path: Path) -> bool:
    try:
        run_git(path, ["rev-parse", "--git-dir"])
        return True
    except GitError:
        return False


def get_diff_text(repo_root: Path, diff_range: str) -> str:
    _validate_diff_range(diff_range)
    return run_git(repo_root, ["diff", *_SAFE_DIFF_FLAGS, diff_range])


def _parse_hunk_header(match: re.Match[str], path: Path) -> DiffHunk:
    old_start = int(match.group(1))
    old_len_str = match.group(2)
    old_len = int(old_len_str) if old_len_str else 1
    new_start = int(match.group(3))
    new_len_str = match.group(4)
    new_len = int(new_len_str) if new_len_str else 1

    return DiffHunk(
        path=path,
        new_start=new_start,
        new_len=new_len,
        old_start=old_start,
        old_len=old_len,
    )


_C_ESCAPE_MAP = {"t": "\t", "n": "\n", "r": "\r", "b": "\b", "f": "\f", "v": "\v", "a": "\a", "\\": "\\", '"': '"'}


def _unquote_c_style(quoted: str) -> str:
    if not (quoted.startswith('"') and quoted.endswith('"')):
        return quoted
    raw = quoted[1:-1]
    chars: list[str] = []
    i = 0
    while i < len(raw):
        if raw[i] == "\\" and i + 1 < len(raw):
            nxt = raw[i + 1]
            if nxt in _C_ESCAPE_MAP:
                chars.append(_C_ESCAPE_MAP[nxt])
                i += 2
            elif nxt in "01234567" and i + 3 <= len(raw) and all(c in "01234567" for c in raw[i + 1 : i + 4]):
                chars.append(chr(int(raw[i + 1 : i + 4], 8)))
                i += 4
            else:
                chars.append("\\")
                i += 1
        else:
            chars.append(raw[i])
            i += 1
    result = "".join(chars)
    try:
        return result.encode("latin-1").decode("utf-8")
    except (UnicodeDecodeError, UnicodeEncodeError):
        return result


def _parse_path_line(line: str, repo_root: Path) -> tuple[str, Path | None]:
    if line.startswith("--- /dev/null"):
        return "old", None
    if line.startswith("+++ /dev/null"):
        return "new", None
    if line.startswith("--- a/"):
        rel_path = line.removeprefix("--- a/").strip()
        resolved = (repo_root / rel_path).resolve()
        if not resolved.is_relative_to(repo_root.resolve()):
            return "", None
        return "old", repo_root / rel_path
    if line.startswith("+++ b/"):
        rel_path = line.removeprefix("+++ b/").strip()
        resolved = (repo_root / rel_path).resolve()
        if not resolved.is_relative_to(repo_root.resolve()):
            return "", None
        return "new", repo_root / rel_path
    if line.startswith('--- "a/'):
        rel_path = _unquote_c_style(line.removeprefix("--- ").strip()).removeprefix("a/")
        resolved = (repo_root / rel_path).resolve()
        if not resolved.is_relative_to(repo_root.resolve()):
            return "", None
        return "old", repo_root / rel_path
    if line.startswith('+++ "b/'):
        rel_path = _unquote_c_style(line.removeprefix("+++ ").strip()).removeprefix("b/")
        resolved = (repo_root / rel_path).resolve()
        if not resolved.is_relative_to(repo_root.resolve()):
            return "", None
        return "new", repo_root / rel_path
    return "", None


def parse_diff(repo_root: Path, diff_range: str) -> list[DiffHunk]:
    _validate_diff_range(diff_range)
    output = run_git(repo_root, ["diff", *_SAFE_DIFF_FLAGS, "--unified=0", "-M", diff_range])
    hunks: list[DiffHunk] = []
    old_path: Path | None = None
    new_path: Path | None = None

    for line in output.splitlines():
        path_type, path = _parse_path_line(line, repo_root)
        if path_type == "old":
            old_path = path
            continue
        if path_type == "new":
            new_path = path
            continue

        match = _HUNK_RE.match(line)
        if match:
            current_path = new_path if new_path else old_path
            if current_path:
                hunks.append(_parse_hunk_header(match, current_path))

    return hunks


def _run_git_z(repo_root: Path, args: list[str]) -> list[str]:
    output = run_git(repo_root, args)
    return [p for p in output.split("\0") if p]


def get_changed_files(repo_root: Path, diff_range: str) -> list[Path]:
    _validate_diff_range(diff_range)
    return [repo_root / p for p in _run_git_z(repo_root, ["diff", *_SAFE_DIFF_FLAGS, "--name-only", "-M", "-z", diff_range])]


def split_diff_range(diff_range: str) -> tuple[str | None, str | None]:
    m = _RANGE_RE.match(diff_range)
    if not m:
        return (None, None)
    base = m.group(1).strip()
    head = m.group(3).strip()
    return (base or None, head or None)


def get_untracked_files(repo_root: Path) -> list[Path]:
    return [repo_root / p for p in _run_git_z(repo_root, ["ls-files", "--others", "--exclude-standard", "-z"])]


def get_deleted_files(repo_root: Path, diff_range: str) -> set[Path]:
    _validate_diff_range(diff_range)
    return {
        (repo_root / p).resolve()
        for p in _run_git_z(repo_root, ["diff", *_SAFE_DIFF_FLAGS, "--diff-filter=D", "--name-only", "-M", "-z", diff_range])
    }


def get_renamed_paths(repo_root: Path, diff_range: str, min_similarity: int = 95) -> tuple[set[Path], set[Path]]:
    _validate_diff_range(diff_range)
    output = run_git(repo_root, ["diff", *_SAFE_DIFF_FLAGS, "--diff-filter=R", "--name-status", "-M", "-z", diff_range])
    parts = output.split("\0")
    old_paths: set[Path] = set()
    pure_new_paths: set[Path] = set()
    i = 0
    while i < len(parts):
        if parts[i].startswith("R"):
            try:
                sim = int(parts[i][1:])
            except ValueError:
                sim = 0
            if i + 1 < len(parts) and parts[i + 1]:
                old_paths.add((repo_root / parts[i + 1]).resolve())
            if sim >= min_similarity and i + 2 < len(parts) and parts[i + 2]:
                pure_new_paths.add((repo_root / parts[i + 2]).resolve())
            i += 3
        else:
            i += 1
    return old_paths, pure_new_paths


def show_file_at_revision(repo_root: Path, rev: str, rel_path: Path) -> str:
    spec = f"{rev}:{rel_path.as_posix()}"
    return run_git(repo_root, ["show", spec])


class CatFileBatch:
    def __init__(self, repo_root: Path) -> None:
        self._proc: subprocess.Popen[bytes] | None = None
        self._repo_root = repo_root

    def _ensure_started(self) -> subprocess.Popen[bytes]:
        if self._proc is None or self._proc.poll() is not None:
            self._proc = subprocess.Popen(
                ["git", "-C", str(self._repo_root), "cat-file", "--batch"],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
            )
        return self._proc

    def get(self, rev: str, rel_path: Path) -> str:
        spec = f"{rev}:{rel_path.as_posix()}\n"
        proc = self._ensure_started()
        assert proc.stdin is not None and proc.stdout is not None
        proc.stdin.write(spec.encode())
        proc.stdin.flush()

        header = proc.stdout.readline()
        if not header:
            raise GitError(f"cat-file: unexpected EOF for {spec.strip()}")

        header_str = header.decode("utf-8", errors="replace").strip()
        if header_str.endswith("missing"):
            raise GitError(f"Path not found: {spec.strip()}")

        parts = header_str.split()
        if len(parts) < 3:
            raise GitError(f"cat-file: malformed header: {header_str}")

        size = int(parts[2])
        content = proc.stdout.read(size)
        proc.stdout.read(1)  # trailing LF

        return content.decode("utf-8", errors="replace")

    def close(self) -> None:
        if self._proc is not None and self._proc.poll() is None:
            assert self._proc.stdin is not None
            self._proc.stdin.close()
            try:
                self._proc.wait(timeout=5)
            except subprocess.TimeoutExpired:
                self._proc.kill()
                self._proc.wait()
            self._proc = None

    def __enter__(self) -> CatFileBatch:
        return self

    def __exit__(
        self,
        _exc_type: type[BaseException] | None,
        _exc_val: BaseException | None,
        _exc_tb: TracebackType | None,
    ) -> None:
        self.close()

File: diffctx/src/git.rs
use std::collections::HashSet;
use std::io::{BufRead, BufReader, Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Child, ChildStdout, Command, Stdio};
use std::sync::Arc;
use std::time::Duration;

use once_cell::sync::Lazy;
use regex::Regex;

use crate::types::DiffHunk;

const GIT_TIMEOUT_SECS: u64 = 60;
const SAFE_DIFF_FLAGS: &[&str] = &["--no-textconv", "--no-ext-diff"];

static HUNK_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@").unwrap());

static RANGE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^\s*(\S+?)(\.\.\.?)(\S*?)\s*$").unwrap());

static SAFE_RANGE_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^[a-zA-Z0-9_.^~/@{}\-]+(\.\.\.?[a-zA-Z0-9_.^~/@{}\-]*)?$").unwrap());

#[derive(Debug, thiserror::Error)]
pub enum GitError {
    #[error("git command failed: {0}")]
    CommandFailed(String),
    #[error("not a git repository: {0}")]
    NotARepo(PathBuf),
    #[error("invalid diff range: {0}")]
    InvalidRange(String),
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("timeout after {0}s")]
    Timeout(u64),
}

pub type Result<T> = std::result::Result<T, GitError>;

fn validate_diff_range(diff_range: &str) -> Result<()> {
    if !SAFE_RANGE_RE.is_match(diff_range.trim()) {
        return Err(GitError::InvalidRange(diff_range.to_string()));
    }
    Ok(())
}

pub fn run_git(repo_root: &Path, args: &[&str]) -> Result<String> {
    let mut cmd = Command::new("git");
    cmd.arg("-C")
        .arg(repo_root)
        .args(args)
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());

    let child = cmd.spawn().map_err(|e| {
        if e.kind() == std::io::ErrorKind::NotFound {
            GitError::CommandFailed("git is not installed or not in PATH".into())
        } else {
            GitError::Io(e)
        }
    })?;

    let output = wait_with_timeout(child, Duration::from_secs(GIT_TIMEOUT_SECS), args)?;

    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(GitError::CommandFailed(format!(
            "git {} failed: {}",
            args.join(" "),
            stderr.trim()
        )));
    }

    Ok(String::from_utf8_lossy(&output.stdout).into_owned())
}

fn wait_with_timeout(
    child: Child,
    timeout: Duration,
    _args: &[&str],
) -> Result<std::process::Output> {
    let mut child = child;
    let start = std::time::Instant::now();

    loop {
        match child.try_wait() {
            Ok(Some(status)) => {
                let mut stdout = Vec::new();
                let mut stderr = Vec::new();
                if let Some(mut out) = child.stdout.take() {
                    let _ = out.read_to_end(&mut stdout);
                }
                if let Some(mut err) = child.stderr.take() {
                    let _ = err.read_to_end(&mut stderr);
                }
                return Ok(std::process::Output {
                    status,
                    stdout,
                    stderr,
                });
            }
            Ok(None) => {
                if start.elapsed() >= timeout {
                    let _ = child.kill();
                    let _ = child.wait();
                    return Err(GitError::Timeout(timeout.as_secs()));
                }
                std::thread::sleep(Duration::from_millis(10));
            }
            Err(e) => return Err(GitError::Io(e)),
        }
    }
}

pub fn is_git_repo(path: &Path) -> bool {
    run_git(path, &["rev-parse", "--git-dir"]).is_ok()
}

pub fn get_diff_text(repo_root: &Path, diff_range: Option<&str>) -> Result<String> {
    let mut args: Vec<&str> = vec!["diff"];
    args.extend_from_slice(SAFE_DIFF_FLAGS);
    if let Some(range) = diff_range {
        validate_diff_range(range)?;
        args.push(range);
    }
    run_git(repo_root, &args)
}

fn unquote_c_style(quoted: &str) -> String {
    if !(quoted.starts_with('"') && quoted.ends_with('"')) {
        return quoted.to_string();
    }

    let raw = &quoted[1..quoted.len() - 1];
    let bytes = raw.as_bytes();
    let mut result: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut i = 0;

    while i < bytes.len() {
        if bytes[i] == b'\\' && i + 1 < bytes.len() {
            let nxt = bytes[i + 1];
            match nxt {
                b't' => { result.push(b'\t'); i += 2; }
                b'n' => { result.push(b'\n'); i += 2; }
                b'r' => { result.push(b'\r'); i += 2; }
                b'b' => { result.push(0x08); i += 2; }
                b'f' => { result.push(0x0C); i += 2; }
                b'v' => { result.push(0x0B); i += 2; }
                b'a' => { result.push(0x07); i += 2; }
                b'\\' => { result.push(b'\\'); i += 2; }
                b'"' => { result.push(b'"'); i += 2; }
                b'0'..=b'7' if i + 3 < bytes.len()
                    && bytes[i + 2].is_ascii_digit() && bytes[i + 2] <= b'7'
                    && bytes[i + 3].is_ascii_digit() && bytes[i + 3] <= b'7' =>
                {
                    let val = (nxt - b'0') * 64
                        + (bytes[i + 2] - b'0') * 8
                        + (bytes[i + 3] - b'0');
                    result.push(val);
                    i += 4;
                }
                _ => {
                    result.push(b'\\');
                    i += 1;
                }
            }
        } else {
            result.push(bytes[i]);
            i += 1;
        }
    }

    String::from_utf8(result).unwrap_or_else(|e| String::from_utf8_lossy(e.as_bytes()).into_owned())
}

fn parse_path_line(line: &str, repo_root: &Path) -> (&'static str, Option<PathBuf>) {
    let resolved_root = repo_root
        .canonicalize()
        .unwrap_or_else(|_| repo_root.to_path_buf());

    if line.starts_with("--- /dev/null") {
        return ("old", None);
    }
    if line.starts_with("+++ /dev/null") {
        return ("new", None);
    }

    if let Some(rest) = line.strip_prefix("--- a/") {
        let rel_path = rest.trim();
        let resolved = (repo_root.join(rel_path))
            .canonicalize()
            .unwrap_or_else(|_| repo_root.join(rel_path));
        if !resolved.starts_with(&resolved_root) {
            return ("", None);
        }
        return ("old", Some(repo_root.join(rel_path)));
    }

    if let Some(rest) = line.strip_prefix("+++ b/") {
        let rel_path = rest.trim();
        let resolved = (repo_root.join(rel_path))
            .canonicalize()
            .unwrap_or_else(|_| repo_root.join(rel_path));
        if !resolved.starts_with(&resolved_root) {
            return ("", None);
        }
        return ("new", Some(repo_root.join(rel_path)));
    }

    if line.starts_with("--- \"a/") {
        let quoted = line.strip_prefix("--- ").unwrap().trim();
        let unquoted = unquote_c_style(quoted);
        let rel_path = unquoted.strip_prefix("a/").unwrap_or(&unquoted);
        let resolved = (repo_root.join(rel_path))
            .canonicalize()
            .unwrap_or_else(|_| repo_root.join(rel_path));
        if !resolved.starts_with(&resolved_root) {
            return ("", None);
        }
        return ("old", Some(repo_root.join(rel_path)));
    }

    if line.starts_with("+++ \"b/") {
        let quoted = line.strip_prefix("+++ ").unwrap().trim();
        let unquoted = unquote_c_style(quoted);
        let rel_path = unquoted.strip_prefix("b/").unwrap_or(&unquoted);
        let resolved = (repo_root.join(rel_path))
            .canonicalize()
            .unwrap_or_else(|_| repo_root.join(rel_path));
        if !resolved.starts_with(&resolved_root) {
            return ("", None);
        }
        return ("new", Some(repo_root.join(rel_path)));
    }

    ("", None)
}

fn parse_hunk_header(caps: &regex::Captures, path: &Path) -> DiffHunk {
    let old_start: u32 = caps[1].parse().unwrap();
    let old_len: u32 = caps.get(2).map_or(1, |m| m.as_str().parse().unwrap());
    let new_start: u32 = caps[3].parse().unwrap();
    let new_len: u32 = caps.get(4).map_or(1, |m| m.as_str().parse().unwrap());

    DiffHunk {
        path: Arc::from(path.to_string_lossy().as_ref()),
        new_start,
        new_len,
        old_start,
        old_len,
    }
}

pub fn parse_diff(repo_root: &Path, diff_range: Option<&str>) -> Result<Vec<DiffHunk>> {
    let mut args: Vec<&str> = vec!["diff"];
    args.extend_from_slice(SAFE_DIFF_FLAGS);
    args.push("--unified=0");
    args.push("-M");
    if let Some(range) = diff_range {
        validate_diff_range(range)?;
        args.push(range);
    }

    let output = run_git(repo_root, &args)?;
    let mut hunks = Vec::new();
    let mut old_path: Option<PathBuf> = None;
    let mut new_path: Option<PathBuf> = None;

    for line in output.lines() {
        let (path_type, path) = parse_path_line(line, repo_root);
        match path_type {
            "old" => {
                old_path = path;
                continue;
            }
            "new" => {
                new_path = path;
                continue;
            }
            _ => {}
        }

        if let Some(caps) = HUNK_RE.captures(line) {
            let current_path = new_path.as_deref().or(old_path.as_deref());
            if let Some(p) = current_path {
                hunks.push(parse_hunk_header(&caps, p));
            }
        }
    }

    Ok(hunks)
}

fn run_git_z(repo_root: &Path, args: &[&str]) -> Result<Vec<String>> {
    let output = run_git(repo_root, args)?;
    Ok(output
        .split('\0')
        .filter(|s| !s.is_empty())
        .map(String::from)
        .collect())
}

pub fn get_changed_files(repo_root: &Path, diff_range: Option<&str>) -> Result<Vec<PathBuf>> {
    let mut args: Vec<&str> = vec!["diff"];
    args.extend_from_slice(SAFE_DIFF_FLAGS);
    args.extend_from_slice(&["--name-only", "-M", "-z"]);
    if let Some(range) = diff_range {
        validate_diff_range(range)?;
        args.push(range);
    }
    let parts = run_git_z(repo_root, &args)?;
    Ok(parts.iter().map(|p| repo_root.join(p)).collect())
}

pub fn get_deleted_files(
    repo_root: &Path,
    diff_range: Option<&str>,
) -> Result<HashSet<PathBuf>> {
    let mut args: Vec<&str> = vec!["diff"];
    args.extend_from_slice(SAFE_DIFF_FLAGS);
    args.extend_from_slice(&["--diff-filter=D", "--name-only", "-M", "-z"]);
    if let Some(range) = diff_range {
        validate_diff_range(range)?;
        args.push(range);
    }
    let parts = run_git_z(repo_root, &args)?;
    Ok(parts
        .iter()
        .map(|p| {
            repo_root
                .join(p)
                .canonicalize()
                .unwrap_or_else(|_| repo_root.join(p))
        })
        .collect())
}

pub fn get_renamed_paths(
    repo_root: &Path,
    diff_range: Option<&str>,
    min_similarity: u32,
) -> Result<(HashSet<PathBuf>, HashSet<PathBuf>)> {
    let mut args: Vec<&str> = vec!["diff"];
    args.extend_from_slice(SAFE_DIFF_FLAGS);
    args.extend_from_slice(&["--diff-filter=R", "--name-status", "-M", "-z"]);
    if let Some(range) = diff_range {
        validate_diff_range(range)?;
        args.push(range);
    }
    let output = run_git(repo_root, &args)?;
    let parts: Vec<&str> = output.split('\0').collect();

    let mut old_paths = HashSet::new();
    let mut pure_new_paths = HashSet::new();
    let mut i = 0;

    while i < parts.len() {
        if parts[i].starts_with('R') {
            let sim: u32 = parts[i][1..].parse().unwrap_or(0);

            if i + 1 < parts.len() && !parts[i + 1].is_empty() {
                let resolved = repo_root
                    .join(parts[i + 1])
                    .canonicalize()
                    .unwrap_or_else(|_| repo_root.join(parts[i + 1]));
                old_paths.insert(resolved);
            }

            if sim >= min_similarity && i + 2 < parts.len() && !parts[i + 2].is_empty() {
                let resolved = repo_root
                    .join(parts[i + 2])
                    .canonicalize()
                    .unwrap_or_else(|_| repo_root.join(parts[i + 2]));
                pure_new_paths.insert(resolved);
            }

            i += 3;
        } else {
            i += 1;
        }
    }

    Ok((old_paths, pure_new_paths))
}

pub fn split_diff_range(range: &str) -> (Option<String>, Option<String>) {
    match RANGE_RE.captures(range) {
        None => (None, None),
        Some(caps) => {
            let base = caps.get(1).map(|m| m.as_str().trim().to_string()).filter(|s| !s.is_empty());
            let head = caps.get(3).map(|m| m.as_str().trim().to_string()).filter(|s| !s.is_empty());
            (base, head)
        }
    }
}

pub fn show_file_at_revision(repo_root: &Path, rev: &str, rel_path: &Path) -> Result<String> {
    let spec = format!("{}:{}", rev, rel_path.to_string_lossy().replace('\\', "/"));
    run_git(repo_root, &["show", &spec])
}

pub fn get_commit_message(repo_root: &Path, rev: &str) -> Result<String> {
    match run_git(repo_root, &["log", "-1", "--format=%s%n%b", rev]) {
        Ok(s) => Ok(s.trim().to_string()),
        Err(_) => Ok(String::new()),
    }
}

pub fn get_untracked_files(repo_root: &Path) -> Result<Vec<PathBuf>> {
    let parts = run_git_z(
        repo_root,
        &["ls-files", "--others", "--exclude-standard", "-z"],
    )?;
    Ok(parts.iter().map(|p| repo_root.join(p)).collect())
}

pub struct CatFileBatch {
    repo_root: PathBuf,
    child: Option<Child>,
    reader: Option<BufReader<ChildStdout>>,
}

impl CatFileBatch {
    pub fn new(repo_root: &Path) -> Result<Self> {
        let mut batch = Self {
            repo_root: repo_root.to_path_buf(),
            child: None,
            reader: None,
        };
        batch.ensure_started()?;
        Ok(batch)
    }

    fn ensure_started(&mut self) -> Result<()> {
        let needs_restart = match &mut self.child {
            None => true,
            Some(child) => child.try_wait().ok().flatten().is_some(),
        };

        if needs_restart {
            let mut child = Command::new("git")
                .arg("-C")
                .arg(&self.repo_root)
                .args(["cat-file", "--batch"])
                .stdin(Stdio::piped())
                .stdout(Stdio::piped())
                .stderr(Stdio::null())
                .spawn()?;
            let stdout = child.stdout.take().unwrap();
            self.reader = Some(BufReader::new(stdout));
            self.child = Some(child);
        }

        Ok(())
    }

    pub fn get(&mut self, rev: &str, rel_path: &Path) -> Result<String> {
        let spec = format!(
            "{}:{}\n",
            rev,
            rel_path.to_string_lossy().replace('\\', "/")
        );

        self.ensure_started()?;

        let stdin = self
            .child
            .as_mut()
            .and_then(|c| c.stdin.as_mut())
            .ok_or_else(|| GitError::CommandFailed("cat-file stdin unavailable".into()))?;
        stdin.write_all(spec.as_bytes())?;
        stdin.flush()?;

        let reader = self
            .reader
            .as_mut()
            .ok_or_else(|| GitError::CommandFailed("cat-file stdout unavailable".into()))?;

        let mut header_line = String::new();
        reader.read_line(&mut header_line)?;

        if header_line.is_empty() {
            return Err(GitError::CommandFailed(format!(
                "cat-file: unexpected EOF for {}",
                spec.trim()
            )));
        }

        let header_str = header_line.trim();
        if header_str.ends_with("missing") {
            return Err(GitError::CommandFailed(format!(
                "Path not found: {}",
                spec.trim()
            )));
        }

        let parts: Vec<&str> = header_str.split_whitespace().collect();
        if parts.len() < 3 {
            return Err(GitError::CommandFailed(format!(
                "cat-file: malformed header: {}",
                header_str
            )));
        }

        let size: usize = parts[2].parse().map_err(|_| {
            GitError::CommandFailed(format!("cat-file: invalid size in header: {}", header_str))
        })?;

        let mut content = vec![0u8; size];
        reader.read_exact(&mut content)?;

        let mut trailing = [0u8; 1];
        let _ = reader.read_exact(&mut trailing);

        Ok(String::from_utf8_lossy(&content).into_owned())
    }

    pub fn close(&mut self) {
        self.reader.take();
        if let Some(mut child) = self.child.take() {
            drop(child.stdin.take());
            match child.wait_timeout(Duration::from_secs(5)) {
                Ok(_) => {}
                Err(_) => {
                    let _ = child.kill();
                    let _ = child.wait();
                }
            }
        }
    }
}

impl Drop for CatFileBatch {
    fn drop(&mut self) {
        self.close();
    }
}

trait WaitTimeout {
    fn wait_timeout(&mut self, dur: Duration) -> std::result::Result<std::process::ExitStatus, ()>;
}

impl WaitTimeout for Child {
    fn wait_timeout(&mut self, dur: Duration) -> std::result::Result<std::process::ExitStatus, ()> {
        let start = std::time::Instant::now();
        loop {
            match self.try_wait() {
                Ok(Some(status)) => return Ok(status),
                Ok(None) => {
                    if start.elapsed() >= dur {
                        return Err(());
                    }
                    std::thread::sleep(Duration::from_millis(10));
                }
                Err(_) => return Err(()),
            }
        }
    }
}
